File: go/indicators/jeanphilippepoton/fractalgraphdimensionindex/fractalgraphdimensionindex.go
package fractalgeneralizeddimensionindex

import (
	"fmt"
	"math"
	"sync"

	"zpano/entities"
	"zpano/indicators/core"
	"zpano/indicators/core/outputs"
)

// FractalGraphDimensionIndex computes the Fractal Graph Dimension Index (FGDI).
//
// This is Poton's corrected and enhanced version of the Fractal Dimension
// Index (FDI). It fixes loop boundary and denominator bugs in the original
// and adds standard deviation bands around the estimated dimension.
//
// The indicator is not primed during the first `period - 1` updates.
type FractalGraphDimensionIndex struct {
	mu sync.RWMutex
	core.LineIndicator
	barFunc   entities.BarFunc
	quoteFunc entities.QuoteFunc
	tradeFunc entities.TradeFunc
	window    []float64
	period    int
	nMinus1   int
	winCount  int
	primed    bool
	log2N1    float64
	ln2       float64
	invNSq    float64
	fgdi      float64
	upper     float64
	lower     float64
	stddev    float64
}

// NewFractalGraphDimensionIndex returns an instance of the indicator created using supplied parameters.
func NewFractalGraphDimensionIndex(p *Params) (*FractalGraphDimensionIndex, error) {
	const (
		invalid = "invalid fractal graph dimension index parameters"
		fmts    = "%s: %s"
		fmtw    = "%s: %w"
		fmtn    = "fgdi(%d%s)"
		minper  = 2
	)

	period := p.Period
	if period < minper {
		return nil, fmt.Errorf(fmts, invalid, "period should be greater than 1")
	}

	bc := p.BarComponent
	if bc == 0 {
		bc = entities.DefaultBarComponent
	}

	qc := p.QuoteComponent
	if qc == 0 {
		qc = entities.DefaultQuoteComponent
	}

	tc := p.TradeComponent
	if tc == 0 {
		tc = entities.DefaultTradeComponent
	}

	var (
		err       error
		barFunc   entities.BarFunc
		quoteFunc entities.QuoteFunc
		tradeFunc entities.TradeFunc
	)

	if barFunc, err = entities.BarComponentFunc(bc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	if quoteFunc, err = entities.QuoteComponentFunc(qc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	if tradeFunc, err = entities.TradeComponentFunc(tc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	mnemonic := fmt.Sprintf(fmtn, period, core.ComponentTripleMnemonic(bc, qc, tc))
	desc := "Fractal graph dimension index " + mnemonic

	nMinus1 := period - 1

	ind := &FractalGraphDimensionIndex{
		barFunc:   barFunc,
		quoteFunc: quoteFunc,
		tradeFunc: tradeFunc,
		window:    make([]float64, period),
		period:    period,
		nMinus1:   nMinus1,
		log2N1:    math.Log(2.0 * float64(nMinus1)),
		ln2:       math.Log(2.0),
		invNSq:    1.0 / float64(period*period),
		fgdi:       math.NaN(),
		upper:     math.NaN(),
		lower:     math.NaN(),
		stddev:    math.NaN(),
	}

	ind.LineIndicator = core.NewLineIndicator(mnemonic, desc, barFunc, quoteFunc, tradeFunc, ind.Update)

	return ind, nil
}

// IsPrimed indicates whether the indicator is primed.
func (f *FractalGraphDimensionIndex) IsPrimed() bool {
	f.mu.RLock()
	defer f.mu.RUnlock()

	return f.primed
}

// Metadata describes the output data of the indicator.
func (f *FractalGraphDimensionIndex) Metadata() core.Metadata {
	return core.BuildMetadata(
		core.FractalGraphDimensionIndex,
		f.LineIndicator.Mnemonic,
		f.LineIndicator.Description,
		[]core.OutputText{
			{Mnemonic: f.LineIndicator.Mnemonic, Description: f.LineIndicator.Description},
			{Mnemonic: f.LineIndicator.Mnemonic + " upper", Description: f.LineIndicator.Description + " Upper"},
			{Mnemonic: f.LineIndicator.Mnemonic + " lower", Description: f.LineIndicator.Description + " Lower"},
			{Mnemonic: f.LineIndicator.Mnemonic + " stddev", Description: f.LineIndicator.Description + " Stddev"},
			{Mnemonic: f.LineIndicator.Mnemonic + " band", Description: f.LineIndicator.Description + " Band"},
		},
	)
}

// Update updates the value of the fractal graph dimension index given the next sample.
// Returns the FGDI value. Use Fgdi(), Upper(), Lower(), Stddev() for all outputs.
func (f *FractalGraphDimensionIndex) Update(sample float64) float64 {
	if math.IsNaN(sample) {
		return sample
	}

	f.mu.Lock()
	defer f.mu.Unlock()

	period := f.period
	nMinus1 := f.nMinus1

	if f.primed {
		for i := 0; i < nMinus1; i++ {
			f.window[i] = f.window[i+1]
		}

		f.window[nMinus1] = sample
	} else {
		f.window[f.winCount] = sample
		f.winCount++

		if f.winCount < period {
			return math.NaN()
		}

		f.primed = true
	}

	// Find min/max for normalization.
	priceMax := f.window[0]
	priceMin := f.window[0]

	for k := 1; k < period; k++ {
		if f.window[k] > priceMax {
			priceMax = f.window[k]
		}

		if f.window[k] < priceMin {
			priceMin = f.window[k]
		}
	}

	priceRange := priceMax - priceMin
	if priceRange < 1e-10 {
		f.fgdi = 1.0
		f.stddev = 0.0
		f.upper = 1.0
		f.lower = 1.0

		return 1.0
	}

	// Normalize and compute path segments.
	priorNorm := (f.window[0] - priceMin) / priceRange
	length := 0.0
	segments := make([]float64, nMinus1)

	for k := 1; k < period; k++ {
		currNorm := (f.window[k] - priceMin) / priceRange
		diff := currNorm - priorNorm
		seg := math.Sqrt(diff*diff + f.invNSq)
		segments[k-1] = seg
		length += seg
		priorNorm = currNorm
	}

	// FGDI = 1 + (ln(L) + ln(2)) / ln(2*(N-1))
	fgdi := 1.0 + (math.Log(length)+f.ln2)/f.log2N1

	// Standard deviation of the estimate.
	meanSeg := length / float64(nMinus1)
	sumSq := 0.0

	for k := 0; k < nMinus1; k++ {
		d := segments[k] - meanSeg
		sumSq += d * d
	}

	variance := sumSq / (length * length * f.log2N1 * f.log2N1)
	stddev := math.Sqrt(variance)

	f.fgdi = fgdi
	f.upper = fgdi + stddev
	f.lower = fgdi - stddev
	f.stddev = stddev

	return fgdi
}

// UpdateAll updates the indicator and returns all four outputs: fgdi, upper, lower, stddev.
func (f *FractalGraphDimensionIndex) UpdateAll(sample float64) (float64, float64, float64, float64) {
	fgdi := f.Update(sample)

	f.mu.RLock()
	defer f.mu.RUnlock()

	return fgdi, f.upper, f.lower, f.stddev
}

// FgdiValue returns the last computed FGDI value.
func (f *FractalGraphDimensionIndex) FgdiValue() float64 {
	f.mu.RLock()
	defer f.mu.RUnlock()

	return f.fgdi
}

// UpperValue returns the last computed upper band value.
func (f *FractalGraphDimensionIndex) UpperValue() float64 {
	f.mu.RLock()
	defer f.mu.RUnlock()

	return f.upper
}

// LowerValue returns the last computed lower band value.
func (f *FractalGraphDimensionIndex) LowerValue() float64 {
	f.mu.RLock()
	defer f.mu.RUnlock()

	return f.lower
}

// StddevValue returns the last computed stddev value.
func (f *FractalGraphDimensionIndex) StddevValue() float64 {
	f.mu.RLock()
	defer f.mu.RUnlock()

	return f.stddev
}

// UpdateScalar updates the indicator given the next scalar sample.
func (f *FractalGraphDimensionIndex) UpdateScalar(sample *entities.Scalar) core.Output {
	fgdi, upper, lower, stddev := f.UpdateAll(sample.Value)

	const outputCount = 5

	output := make([]any, outputCount)
	output[0] = entities.Scalar{Time: sample.Time, Value: fgdi}
	output[1] = entities.Scalar{Time: sample.Time, Value: upper}
	output[2] = entities.Scalar{Time: sample.Time, Value: lower}
	output[3] = entities.Scalar{Time: sample.Time, Value: stddev}

	if math.IsNaN(lower) || math.IsNaN(upper) {
		output[4] = outputs.NewEmptyBand(sample.Time)
	} else {
		output[4] = outputs.NewBand(sample.Time, lower, upper)
	}

	return output
}

// UpdateBar updates the indicator given the next bar sample.
func (f *FractalGraphDimensionIndex) UpdateBar(sample *entities.Bar) core.Output {
	v := f.barFunc(sample)

	return f.UpdateScalar(&entities.Scalar{Time: sample.Time, Value: v})
}

// UpdateQuote updates the indicator given the next quote sample.
func (f *FractalGraphDimensionIndex) UpdateQuote(sample *entities.Quote) core.Output {
	v := f.quoteFunc(sample)

	return f.UpdateScalar(&entities.Scalar{Time: sample.Time, Value: v})
}

// UpdateTrade updates the indicator given the next trade sample.
func (f *FractalGraphDimensionIndex) UpdateTrade(sample *entities.Trade) core.Output {
	v := f.tradeFunc(sample)

	return f.UpdateScalar(&entities.Scalar{Time: sample.Time, Value: v})
}

File: ts/indicators/jean-philippe-poton/fractal-graph-dimension-index/fractal-graph-dimension-index.ts
import { buildMetadata } from '../../core/build-metadata';
import { componentTripleMnemonic } from '../../core/component-triple-mnemonic';
import { IndicatorMetadata } from '../../core/indicator-metadata';
import { IndicatorOutput } from '../../core/indicator-output';
import { IndicatorIdentifier } from '../../core/indicator-identifier';
import { LineIndicator } from '../../core/line-indicator';
import { Band } from '../../core/outputs/band';
import { Scalar } from '../../../entities/scalar';
import { Bar } from '../../../entities/bar';
import { Quote } from '../../../entities/quote';
import { Trade } from '../../../entities/trade';
import { FractalGraphDimensionIndexParams } from './params';

/** Function to calculate mnemonic of a __FractalGraphDimensionIndex__ indicator. */
export const fractalGraphDimensionIndexMnemonic = (params: FractalGraphDimensionIndexParams): string =>
    'fgdi('.concat(params.period.toString(), componentTripleMnemonic(params.barComponent, params.quoteComponent, params.tradeComponent), ')');

/** Fractal Graph Dimension Index indicator. */
export class FractalGraphDimensionIndex extends LineIndicator {
    private window: Array<number>;
    private period: number;
    private windowCount: number;
    private nMinus1: number;
    private log2N1: number;
    private ln2: number;
    private invNSq: number;
    private _fgdi: number;
    private _upper: number;
    private _lower: number;
    private _stddev: number;

    /**
     * Constructs an instance given a period.
     * The period should be an integer greater than 1.
     **/
    public constructor(params: FractalGraphDimensionIndexParams) {
        super();
        const period = Math.floor(params.period);
        if (period < 2) {
            throw new Error('period should be greater than 1');
        }

        this.mnemonic = fractalGraphDimensionIndexMnemonic(params);
        this.description = 'Fractal graph dimension index ' + this.mnemonic;
        this.barComponent = params.barComponent;
        this.quoteComponent = params.quoteComponent;
        this.tradeComponent = params.tradeComponent;
        this.window = new Array<number>(period);
        this.period = period;
        this.nMinus1 = period - 1;
        this.windowCount = 0;
        this.primed = false;
        this.log2N1 = Math.log(2.0 * (period - 1));
        this.ln2 = Math.log(2.0);
        this.invNSq = 1.0 / (period * period);
        this._fgdi = Number.NaN;
        this._upper = Number.NaN;
        this._lower = Number.NaN;
        this._stddev = Number.NaN;
    }

    /** Describes the output data of the indicator. */
    public metadata(): IndicatorMetadata {
        return buildMetadata(
            IndicatorIdentifier.FractalGraphDimensionIndex,
            this.mnemonic,
            this.description,
            [
                { mnemonic: this.mnemonic, description: this.description },
                { mnemonic: this.mnemonic + ' upper', description: this.description + ' Upper' },
                { mnemonic: this.mnemonic + ' lower', description: this.description + ' Lower' },
                { mnemonic: this.mnemonic + ' stddev', description: this.description + ' Stddev' },
                { mnemonic: this.mnemonic + ' band', description: this.description + ' Band' },
            ],
        );
    }

    /** Updates the indicator and returns the FGDI value. Use fgdiValue, upperValue, lowerValue, stddevValue for all outputs. */
    public update(sample: number): number {
        if (Number.isNaN(sample)) {
            return sample;
        }

        const period = this.period;
        const nMinus1 = this.nMinus1;

        if (this.primed) {
            for (let i = 0; i < nMinus1; i++) {
                this.window[i] = this.window[i + 1];
            }

            this.window[nMinus1] = sample;
        } else {
            this.window[this.windowCount] = sample;
            this.windowCount++;

            if (this.windowCount < period) {
                return Number.NaN;
            }

            this.primed = true;
        }

        // Find min/max for normalization.
        let priceMax = this.window[0];
        let priceMin = this.window[0];

        for (let k = 1; k < period; k++) {
            if (this.window[k] > priceMax) { priceMax = this.window[k]; }
            if (this.window[k] < priceMin) { priceMin = this.window[k]; }
        }

        const priceRange = priceMax - priceMin;
        if (priceRange < 1e-10) {
            this._fgdi = 1.0;
            this._stddev = 0.0;
            this._upper = 1.0;
            this._lower = 1.0;
            return 1.0;
        }

        // Normalize and compute path segments.
        let priorNorm = (this.window[0] - priceMin) / priceRange;
        let length = 0.0;
        const segments = new Array<number>(nMinus1);

        for (let k = 1; k < period; k++) {
            const currNorm = (this.window[k] - priceMin) / priceRange;
            const diff = currNorm - priorNorm;
            const seg = Math.sqrt(diff * diff + this.invNSq);
            segments[k - 1] = seg;
            length += seg;
            priorNorm = currNorm;
        }

        // FGDI = 1 + (ln(L) + ln(2)) / ln(2*(N-1))
        const fgdi = 1.0 + (Math.log(length) + this.ln2) / this.log2N1;

        // Standard deviation of the estimate.
        const meanSeg = length / nMinus1;
        let sumSq = 0.0;

        for (let k = 0; k < nMinus1; k++) {
            const d = segments[k] - meanSeg;
            sumSq += d * d;
        }

        const variance = sumSq / (length * length * this.log2N1 * this.log2N1);
        const stddev = Math.sqrt(variance);

        this._fgdi = fgdi;
        this._upper = fgdi + stddev;
        this._lower = fgdi - stddev;
        this._stddev = stddev;

        return fgdi;
    }

    /** Updates and returns all four outputs: [fgdi, upper, lower, stddev]. */
    public updateAll(sample: number): [number, number, number, number] {
        const fgdi = this.update(sample);
        return [fgdi, this._upper, this._lower, this._stddev];
    }

    /** Returns the last computed FGDI value. */
    public get fgdiValue(): number { return this._fgdi; }

    /** Returns the last computed upper band value. */
    public get upperValue(): number { return this._upper; }

    /** Returns the last computed lower band value. */
    public get lowerValue(): number { return this._lower; }

    /** Returns the last computed stddev value. */
    public get stddevValue(): number { return this._stddev; }

    /** Updates the indicator given the next scalar sample. */
    public updateScalar(sample: Scalar): IndicatorOutput {
        const [fgdi, upper, lower, stddev] = this.updateAll(sample.value);

        const s0 = new Scalar(); s0.time = sample.time; s0.value = fgdi;
        const s1 = new Scalar(); s1.time = sample.time; s1.value = upper;
        const s2 = new Scalar(); s2.time = sample.time; s2.value = lower;
        const s3 = new Scalar(); s3.time = sample.time; s3.value = stddev;

        const band = new Band();
        band.time = sample.time;
        if (isNaN(lower) || isNaN(upper)) {
            band.lower = NaN;
            band.upper = NaN;
        } else {
            band.lower = lower;
            band.upper = upper;
        }

        return [s0, s1, s2, s3, band];
    }

    /** Updates the indicator given the next bar sample. */
    public updateBar(sample: Bar): IndicatorOutput {
        const v = this.barComponentFunc(sample);
        const scalar = new Scalar();
        scalar.time = sample.time;
        scalar.value = v;
        return this.updateScalar(scalar);
    }

    /** Updates the indicator given the next quote sample. */
    public updateQuote(sample: Quote): IndicatorOutput {
        const v = this.quoteComponentFunc(sample);
        const scalar = new Scalar();
        scalar.time = sample.time;
        scalar.value = v;
        return this.updateScalar(scalar);
    }

    /** Updates the indicator given the next trade sample. */
    public updateTrade(sample: Trade): IndicatorOutput {
        const v = this.tradeComponentFunc(sample);
        const scalar = new Scalar();
        scalar.time = sample.time;
        scalar.value = v;
        return this.updateScalar(scalar);
    }
}
